In [3]:
import pandas as pd

In [4]:
df = pd.read_csv("/mnt/c/Users/johan/Desktop/Adalab/modulo-3/HR RAW DATA.csv",index_col=0)
df

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField  employeecount  \
0                    6          3               NaN              1   
1                    1          4     Life Sciences              1   
2                    4          2  Technical Degree              1   
3                    2          4           Medical              1   
4                    3          3  Technical Degree              1   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences              1   
1610                 4          2               NaN              1   
1611               -13          5               NaN              1   
1612                 8          4  Technical Degree              1   
1613                 7          2           Medical              1   

     employeenumber  ...  YearsAtCompany  YearsInCurrentRole  \
0             162,0  ...              20                 NaN   
1             259,0  ...              33                 NaN   
2             319,0  ...              22                 NaN   
3               NaN  ...              20                 NaN   
4               NaN  ...              19                 NaN   
...             ...  ...             ...                 ...   
1609          967,0  ...              16                 NaN   
1610          972,0  ...               9                 NaN   
1611            NaN  ...               8                 NaN   
1612          990,0  ...               7                 NaN   
1613            NaN  ...               8                 NaN   

     YearsSinceLastPromotion  YEARSWITHCURRMANAGER  SameAsMonthlyIncome  \
0                         15                    15              19537,0   
1                         11                     9              19999,0   
2                         11                    15              19232,0   
3                          5                     6              17169,0   
4                          2                     8                  NaN   
...                      ...                   ...                  ...   
1609                       3                     7              10325,0   
1610                       0                     8               4447,0   
1611                       0                     7                  NaN   
1612                       0                     7                  NaN   
1613                       0                     7              10845,0   

     DateBirth       Salary                       RoleDepartament  \
0         1972  1000000000$                                   NaN   
1         1971  1000000000$                                   NaN   
2         1981  1000000000$   ManaGER  -  Research & Development    
3         1976  1000000000$                                   NaN   
4         1977  1000000000$                                   NaN   
...        ...          ...                                   ...   
1609      1987  1000000000$                                   NaN   
1610      1978  1000000000$                                   NaN

In [5]:
df_31_42= df.iloc[:, 31:42]

In [6]:
df_31_42

YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0                 20                NaN                       15   
1                 33                NaN                       11   
2                 22                NaN                       11   
3                 20                NaN                        5   
4                 19                NaN                        2   
...              ...                ...                      ...   
1609              16                NaN                        3   
1610               9                NaN                        0   
1611               8                NaN                        0   
1612               7                NaN                        0   
1613               8                NaN                        0   

      YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth       Salary  \
0                       15             19537,0       1972  1000000000$   
1                        9             19999,0       1971  1000000000$   
2                       15             19232,0       1981  1000000000$   
3                        6             17169,0       1976  1000000000$   
4                        8                 NaN       1977  1000000000$   
...                    ...                 ...        ...          ...   
1609                     7             10325,0       1987  1000000000$   
1610                     8              4447,0       1978  1000000000$   
1611                     7                 NaN       1984  1000000000$   
1612                     7                 NaN       1987  1000000000$   
1613                     7             10845,0       1977  1000000000$   

                           RoleDepartament  NUMBERCHILDREN RemoteWork  
0                                      NaN             NaN        Yes  
1                                      NaN             NaN          1  
2      ManaGER  -  Research & Development              NaN          1  
3                                      NaN             NaN      False  
4                                      NaN             NaN          0  
...                                    ...             ...        ...  
1609                                   NaN             NaN          0  
1610                                   NaN             NaN          1  
1611                                   NaN             NaN        Yes  
1612                                   NaN             NaN       True  
1613                                   NaN             NaN          0  

[1614 rows x 10 columns]

In [17]:
df_31_42.isnull()

YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0              False                True                    False   
1              False                True                    False   
2              False                True                    False   
3              False                True                    False   
4              False                True                    False   
...              ...                 ...                      ...   
1609           False                True                    False   
1610           False                True                    False   
1611           False                True                    False   
1612           False                True                    False   
1613           False                True                    False   

      YEARSWITHCURRMANAGER  SameAsMonthlyIncome  DateBirth  Salary  \
0                    False                False      False   False   
1                    False                False      False   False   
2                    False                False      False   False   
3                    False                False      False   False   
4                    False                 True      False   False   
...                    ...                  ...        ...     ...   
1609                 False                False      False   False   
1610                 False                False      False   False   
1611                 False                 True      False   False   
1612                 False                 True      False   False   
1613                 False                False      False   False   

      RoleDepartament  NUMBERCHILDREN  RemoteWork  
0                True            True       False  
1                True            True       False  
2               False            True       False  
3                True            True       False  
4                True            True       False  
...               ...             ...         ...  
1609             True            True       False  
1610             True            True       False  
1611             True            True       False  
1612             True            True       False  
1613             True            True       False  

[1614 rows x 10 columns]

In [18]:
df_31_42.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1614 entries, 0 to 1613
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   YearsAtCompany           1614 non-null   int64  
 1   YearsInCurrentRole       34 non-null     object 
 2   YearsSinceLastPromotion  1614 non-null   int64  
 3   YEARSWITHCURRMANAGER     1614 non-null   int64  
 4   SameAsMonthlyIncome      771 non-null    object 
 5   DateBirth                1614 non-null   int64  
 6   Salary                   1614 non-null   int64  
 7   RoleDepartament          302 non-null    object 
 8   NUMBERCHILDREN           0 non-null      float64
 9   RemoteWork               1614 non-null   bool   
dtypes: bool(1), float64(1), int64(5), object(3)
memory usage: 127.7+ KB


In [19]:
df_31_42.isnull().sum()

YearsAtCompany                0
YearsInCurrentRole         1580
YearsSinceLastPromotion       0
YEARSWITHCURRMANAGER          0
SameAsMonthlyIncome         843
DateBirth                     0
Salary                        0
RoleDepartament            1312
NUMBERCHILDREN             1614
RemoteWork                    0
dtype: int64

In [20]:
df_31_42.duplicated().sum()

147

In [7]:
print(df_31_42.dtypes)

YearsAtCompany               int64
YearsInCurrentRole          object
YearsSinceLastPromotion      int64
YEARSWITHCURRMANAGER         int64
SameAsMonthlyIncome         object
DateBirth                    int64
Salary                      object
RoleDepartament             object
NUMBERCHILDREN             float64
RemoteWork                  object
dtype: object


In [ ]:
YearsAtCompany               int64  
YearsInCurrentRole          object - NaN / Float
YearsSinceLastPromotion      int64
YEARSWITHCURRMANAGER         int64
SameAsMonthlyIncome         object - NaN /Float
DateBirth                    int64
Salary                      object - Dolar $  - Cambiado. 
RoleDepartament             object  - Cambiado a minus. Habia valores mins y mays mezclados. 
NUMBERCHILDREN             float64 - NaN -Eliminada , todos son nulos. 
RemoteWork                  object -  True/ False - Ok cambiado. 

In [8]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'employeecount',
       'employeenumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NUMCOMPANIESWORKED',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TOTALWORKINGYEARS', 'TrainingTimesLastYear', 'WORKLIFEBALANCE',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YEARSWITHCURRMANAGER', 'SameAsMonthlyIncome', 'DateBirth', 'Salary',
       'RoleDepartament', 'NUMBERCHILDREN', 'RemoteWork'],
      dtype='object')

In [26]:
#NUMBERCHILDREN             1614 nulos , la totalidad
#Eliminamos 

df_31_42.drop(['NUMBERCHILDREN'], axis = 1, inplace=True)


In [22]:
#RoleDepartment -Mins

df_31_42['RoleDepartament'] = df_31_42['RoleDepartament'].str.lower()
df_31_42

YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0                 20                NaN                       15   
1                 33                NaN                       11   
2                 22                NaN                       11   
3                 20                NaN                        5   
4                 19                NaN                        2   
...              ...                ...                      ...   
1609              16                NaN                        3   
1610               9                NaN                        0   
1611               8                NaN                        0   
1612               7                NaN                        0   
1613               8                NaN                        0   

      YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth      Salary  \
0                       15             19537,0       1972  1000000000   
1                        9             19999,0       1971  1000000000   
2                       15             19232,0       1981  1000000000   
3                        6             17169,0       1976  1000000000   
4                        8                 NaN       1977  1000000000   
...                    ...                 ...        ...         ...   
1609                     7             10325,0       1987  1000000000   
1610                     8              4447,0       1978  1000000000   
1611                     7                 NaN       1984  1000000000   
1612                     7                 NaN       1987  1000000000   
1613                     7             10845,0       1977  1000000000   

                           RoleDepartament  NUMBERCHILDREN  RemoteWork  
0                                      NaN             NaN        True  
1                                      NaN             NaN        True  
2      manager  -  research & development              NaN        True  
3                                      NaN             NaN        True  
4                                      NaN             NaN       False  
...                                    ...             ...         ...  
1609                                   NaN             NaN       False  
1610                                   NaN             NaN        True  
1611                                   NaN             NaN        True  
1612                                   NaN             NaN        True  
1613                                   NaN             NaN       False  

[1614 rows x 10 columns]

In [ ]:
# Salary - Quitar $
df_31_42['Salary'] = df_31_42['Salary'].str.replace('$', '').astype(int)

In [24]:
df_31_42['Salary']

0       1000000000
1       1000000000
2       1000000000
3       1000000000
4       1000000000
           ...    
1609    1000000000
1610    1000000000
1611    1000000000
1612    1000000000
1613    1000000000
Name: Salary, Length: 1614, dtype: int64

In [11]:
#RemoteWork

df_31_42['RemoteWork'] = df_31_42['RemoteWork'].replace(['yes', '1'], True).replace(['0', 'false'], False)

# Convertir la columna a tipo booleano
df_31_42['RemoteWork'] = df_31_42['RemoteWork'].astype(bool)

In [12]:
df_31_42['RemoteWork']

0        True
1        True
2        True
3        True
4       False
        ...  
1609    False
1610     True
1611     True
1612     True
1613    False
Name: RemoteWork, Length: 1614, dtype: bool

In [13]:
df_31_42['YearsInCurrentRole']

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
1609    NaN
1610    NaN
1611    NaN
1612    NaN
1613    NaN
Name: YearsInCurrentRole, Length: 1614, dtype: object

In [15]:
df_31_42.describe(include = "object").T

count unique                  top freq
YearsInCurrentRole     34     10                  2,0   11
SameAsMonthlyIncome   771    668               6347,0    4
RoleDepartament       302    301   MaNAgeR  -  Sales     2

In [16]:
df_cat = df_31_42.select_dtypes(include = "object")
df_cat

YearsInCurrentRole SameAsMonthlyIncome  \
0                   NaN             19537,0   
1                   NaN             19999,0   
2                   NaN             19232,0   
3                   NaN             17169,0   
4                   NaN                 NaN   
...                 ...                 ...   
1609                NaN             10325,0   
1610                NaN              4447,0   
1611                NaN                 NaN   
1612                NaN                 NaN   
1613                NaN             10845,0   

                           RoleDepartament  
0                                      NaN  
1                                      NaN  
2      ManaGER  -  Research & Development   
3                                      NaN  
4                                      NaN  
...                                    ...  
1609                                   NaN  
1610                                   NaN  
1611                                   NaN  
1612                                   NaN  
1613                                   NaN  

[1614 rows x 3 columns]

In [30]:
# NO FUNCIONA

for elemento in df_31_42.YearsInCurrentRole:
    if type(elemento) == str:
        df_31_42.YearsInCurrentRole = df_31_42.YearsInCurrentRole.replace(',', '.').astype(float)

ValueError: could not convert string to float: '13,0'

In [29]:
df_31_42

YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0                 20                NaN                       15   
1                 33                NaN                       11   
2                 22                NaN                       11   
3                 20                NaN                        5   
4                 19                NaN                        2   
...              ...                ...                      ...   
1609              16                NaN                        3   
1610               9                NaN                        0   
1611               8                NaN                        0   
1612               7                NaN                        0   
1613               8                NaN                        0   

      YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth      Salary  \
0                       15             19537,0       1972  1000000000   
1                        9             19999,0       1971  1000000000   
2                       15             19232,0       1981  1000000000   
3                        6             17169,0       1976  1000000000   
4                        8                 NaN       1977  1000000000   
...                    ...                 ...        ...         ...   
1609                     7             10325,0       1987  1000000000   
1610                     8              4447,0       1978  1000000000   
1611                     7                 NaN       1984  1000000000   
1612                     7                 NaN       1987  1000000000   
1613                     7             10845,0       1977  1000000000   

                           RoleDepartament  RemoteWork  
0                                      NaN        True  
1                                      NaN        True  
2      manager  -  research & development         True  
3                                      NaN        True  
4                                      NaN       False  
...                                    ...         ...  
1609                                   NaN       False  
1610                                   NaN        True  
1611                                   NaN        True  
1612                                   NaN        True  
1613                                   NaN       False  

[1614 rows x 9 columns]

##### Resumen 

* Pendiente qué hacer con las columnas NAN/FLOAT. Se indica que se verá en prox lecciones cómo tratar los nulos. 
* Pendiente cambiar título columnas. mins/ con  _ ?

- YearsAtCompany               int64  
- YearsInCurrentRole          object - NaN / Float - No se puede pasar a INT. 1580 nulos. 
- YearsSinceLastPromotion      int64
- YEARSWITHCURRMANAGER         int64
- SameAsMonthlyIncome         object - NaN /Float - No se puede pasar a INT. 843 nulos.
- DateBirth                    int64
- Salary                      object - Dolar $  - Cambiado.  ✅
- RoleDepartament             object  - Cambiado a minus.✅ Habia valores mins y mays mezclados.  1312 nulos. 
- NUMBERCHILDREN             float64 - NaN -Eliminada , todos son nulos.  ✅
- RemoteWork                  object -  True/ False - Ok cambiado. ✅